gh-132461: Fix crash in OrderedDict.setdefault when key has unstable hash

dura0ok opened the pull request:
Replaced assertion with a runtime check in OrderedDict_setdefault_impl to gracefully handle keys with changing hash values. If such a key is detected, a TypeError is raised instead of aborting the interpreter.


<!-- gh-issue-number: gh-132461 -->
* Issue: gh-132461
<!-- /gh-issue-number -->

dura0ok committed:
Fix crash in OrderedDict.setdefault when key has unstable hash

Replaced assertion with a runtime check in OrderedDict_setdefault_impl
to gracefully handle keys with changing hash values. If such a key is
detected, a TypeError is raised instead of aborting the interpreter.

picnixz commented on Objects/odictobject.c:
```diff
@@ -1033,7 +1033,11 @@ OrderedDict_setdefault_impl(PyODictObject *self, PyObject *key,
                 return NULL;
-            assert(_odict_find_node(self, key) == NULL);
+            if (_odict_find_node(self, key) != NULL) {
+                PyErr_SetString(PyExc_TypeError,
```
We should raise a RuntimeError instead.

dura0ok replied:
yep, fixed, thanks

picnixz reviewed (commented):
Please add a regression test and a NEWS entry as well.

dura0ok committed:
test_ordered_dict: Add test for unstable hash behavior

- Add test_unstable_hash_should_not_abort_issue132461 to ensure that
  OrderedDict handles keys with unstable hashes gracefully.
- Simulates randomized hashes using a custom metaclass and verifies that
  an exception is raised rather than a crash occurring.
- References gh-132461 to track and validate regression coverage.

picnixz commented on Lib/test/test_ordered_dict.py:
```diff
@@ -729,6 +730,22 @@ def test_merge_operator(self):
 
+    def test_unstable_hash_should_not_abort_issue132461(self):
+        # OrderedDict should raise, not abort, when used with unstable-hash keys
+        large_num = 2**64
+
+        class WeirdBase(abc.ABCMeta):
+            def __hash__(self):
+                return randint(0, large_num)
+
+        class weird_bytes(bytes, metaclass=WeirdBase):
+            pass
+
+        obj = self.OrderedDict()
+        with self.assertRaises(Exception):
+            for x in range(100):
+                obj.setdefault(weird_bytes, None)
```
```suggestion
        class A:
            c = 0
            def __hash__(self):
                self.c += 1
                return self.c

        a = A()
        obj = self.OrderedDict()
        msg = "OrderedDict key appears to change its hash value over time"
        with self.assertRaisesRegex(RuntimeError, msg):
            for _ in range(100):
                obj.setdefault(a, None)
```

We can also remove `abc` dependency and `randint` import.

dura0ok replied:
part of test
```
        with self.assertRaises(KeyError):
            for x in range(100):
                obj.setdefault(weird_bytes, None)
```

error
```
ERROR: test_unstable_hash_should_not_abort_issue132461 (test.test_ordered_dict.CPythonOrderedDictTests.test_unstable_hash_should_not_abort_issue132461)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/learning/cpython/Lib/test/test_ordered_dict.py", line 747, in test_unstable_hash_should_not_abort_issue132461
    obj.setdefault(weird_bytes, None)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
RuntimeError: OrderedDict key appears to change its hash value over time
```

fixed part of test
```
        with self.assertRaises(RuntimeError):
            for x in range(100):
                obj.setdefault(weird_bytes, None)
```

```
ERROR: test_unstable_hash_should_not_abort_issue132461 (test.test_ordered_dict.CPythonOrderedDictSubclassTests.test_unstable_hash_should_not_abort_issue132461)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/learning/cpython/Lib/test/test_ordered_dict.py", line 747, in test_unstable_hash_should_not_abort_issue132461
    obj.setdefault(weird_bytes, None)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: <class 'test.test_ordered_dict.OrderedDictTests.test_unstable_hash_should_not_abort_issue132461.<locals>.weird_bytes'>

======================================================================
ERROR: test_unstable_hash_should_not_abort_issue132461 (test.test_ordered_dict.PurePythonOrderedDictSubclassTests.test_unstable_hash_should_not_abort_issue132461)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/learning/cpython/Lib/test/test_ordered_dict.py", line 747, in test_unstable_hash_should_not_abort_issue132461
    obj.setdefault(weird_bytes, None)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/learning/cpython/Lib/collections/__init__.py", line 274, in setdefault
    return self[key]
           ~~~~^^^^^
KeyError: <class 'test.test_ordered_dict.OrderedDictTests.test_unstable_hash_should_not_abort_issue132461.<locals>.weird_bytes'>

======================================================================
ERROR: test_unstable_hash_should_not_abort_issue132461 (test.test_ordered_dict.PurePythonOrderedDictTests.test_unstable_hash_should_not_abort_issue132461)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/learning/cpython/Lib/test/test_ordered_dict.py", line 747, in test_unstable_hash_should_not_abort_issue132461
    obj.setdefault(weird_bytes, None)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/learning/cpython/Lib/collections/__init__.py", line 274, in setdefault
    return self[key]
           ~~~~^^^^^
KeyError: <class 'test.test_ordered_dict.OrderedDictTests.test_unstable_hash_should_not_abort_issue132461.<locals>.weird_bytes'>
```


Very strange

picnixz replied:
Still, we should only catch `RuntimeError` and not arbitrary exceptions. The assertion failing may be an indication that something else is wrong.

sobolevn commented on Lib/test/test_ordered_dict.py:
```diff
@@ -729,6 +730,22 @@ def test_merge_operator(self):
+
+        class WeirdBase(abc.ABCMeta):
+            def __hash__(self):
+                return randint(0, large_num)
```
It would be better to return a new one on each call, otherwise this test will sometimes be flaky, when random will produce the same number twice.
